### DISCUSSION TITLE
Docs: improve guides on context usage

### DISCUSSION THREAD

**Discussion by TkDodo** [pmndrs/zustand#2489] [General]
As [discussed on twitter](https://x.com/dai_shi/status/1779629949806375392), let's think about re-structuring the docs to make some patterns more prominently available, like scoping stores to a component subtree with React Context. I recently wrote [a blogpost](https://tkdodo.eu/blog/zustand-and-react-context) on exactly that topic, and got a lot of questions why this pattern isn't encouraged in the docs.

**Comment by dbritto-dev** [comment]
Let's create a list of all the them:

- createStore + Context API (name TBD)
- slices pattern
- namespaced slices pattern
- persist w/o ssr
- lazy load slices
- react query
- dynamic stores
- async actions
- tutorial
- getting started
- typescript patterns
- react native

and more

**Reply by dai-shi** [reply]
If there are many patterns (and it looks like so), we should probably separate into levels? Like "Basic guides" and "Advanced guides". That way, people who first read those docs don't have to read everything at once.

**Reply by dbritto-dev** [reply]
yeah, I'm down with you

**Comment by dai-shi** [comment]
Apart from "context usage" which this discussion is aiming at, I see the overall structure has room for improvements:
```
getting-started/
guides/
integrations/
migrations/
previous-versions/
```

**Reply by dai-shi** [reply]
So, I'm not sure if we should separate a) overall doc improvements and b) improve the guide for context usage.

**Comment by dai-shi** [comment]
I think @dbritto-dev is the right person to take the lead on improving docs. So, I'd like to nominate him as one of the maintainers for docs.

**Reply by dbritto-dev** [reply]
@dai-shi @TkDodo I've been thinking on moving to `vitepress`,  any thoughts? I'd like to have some features the current page docs doesn't support like copy code, highlight lines, etc

**Reply by dbritto-dev** [reply]
I'm creating a gists with the new structure for zustand docs https://gist.github.com/dbritto-dev/7fc6abc8afb33e5f98df35e6e0aa7657

Looking forward to heard something from you folks cc @TkDodo @dai-shi

**Reply by dai-shi** [reply]
> @dai-shi @TkDodo I've been thinking on moving to `vitepress`, any thoughts? I'd like to have some features the current page docs doesn't support like copy code, highlight lines, etc

Are you saying that we move away from pmndrs/website? I'm fine as long as you continue to maintain it.
There's one requirement from me: We should isolate `docs` from `website` and `docs` should be independent from the website tool. There can be multiple tools. All of zustand, jotai and valtio follow this rule. I'm not sure if vitepress support this pattern.

**Reply by dai-shi** [reply]
> https://gist.github.com/dbritto-dev/7fc6abc8afb33e5f98df35e6e0aa7657

Looks good. I was wondering if we should less suggest `combine` and `subscribeWithSelector` middleware usage and maybe consider dropping it in v6.

**Reply by dbritto-dev** [reply]
@dai-shi I believe we can follow `react-spring` where the project have their own docs website

**Reply by dai-shi** [reply]
I think you misunderstood my requirement. Please see jotai and valtio. It's about separating `docs` and `website` folders in a repository.

**Reply by dbritto-dev** [reply]
> I think you misunderstood my requirement. Please see jotai and valtio. It's about separating `docs` and `website` folders in a repository.

Got it

**Reply by dbritto-dev** [reply]
@dai-shi vitepress can support website and docs structure, ref: https://vitepress.dev/guide/routing. So `root dir` would be `website` and `source dir` would be `docs`

**Reply by dai-shi** [reply]
Cool. But, I wonder if it can really be `./website` and `./docs` in the Zustand repo. Will see. A symbolic link might work.

**Reply by dbritto-dev** [reply]
@dai-shi let me play with that

**Reply by dbritto-dev** [reply]
> Looks good. I was wondering if we should less suggest `combine` and `subscribeWithSelector` middleware usage and maybe consider dropping it in v6.

I'm down with you, we shall mark them as deprecated in v5 and drop them in v6

**Reply by dbritto-dev** [reply]
@TkDodo any thoughts? -> https://gist.github.com/dbritto-dev/7fc6abc8afb33e5f98df35e6e0aa7657

**Reply by dbritto-dev** [reply]
@dai-shi btw, I updated the gist -> https://gist.github.com/dbritto-dev/7fc6abc8afb33e5f98df35e6e0aa7657

**Reply by dai-shi** [reply]
Nice. How about third-party libraries? I guess there can be curated "semi-official" third-party libs. (See this for example:https://github.com/jotaijs/links )

**Reply by dbritto-dev** [reply]
@dai-shi added under Community

**Reply by dai-shi** [reply]
Thanks. btw, "zustand labs" don't mean "semi official". they include both. In Jotai, what's in docs means "semi official". We don't need to follow Jotai way, but zustand labs include both.

**Reply by dbritto-dev** [reply]
@dai-shi I thought the ones under zustand labs are the semi official ones and anything else are the non-official ones

**Reply by dai-shi** [reply]
That's not correct. For example, "zustand-signal" is just an experimental library. Zustand Labs are basically to grow ecosystem and the visibility. So, not all of them are semi official. But, semi-official libs should be in zustand labs.

For example, you can create a new repo in zustand labs for any experiment, but that doesn't automatically mean semi official. It depends, right?

**Reply by dbritto-dev** [reply]
@dai-shi got it, renamed it to just "Third-party Libraries"

**Reply by dbritto-dev** [reply]
@dai-shi btw, the new docs should be available in v4 or v5?

**Reply by dai-shi** [reply]
v4 `main`.

**Reply by TkDodo** [reply]
looks really good 👏 . I would probably put the sections inside `Managing State` that reference middlewares into `Middlewares`. If you read the docs front-to-back, it's probably better to not get introduced to middlewares before the actual `Using middlewares` section.

**Reply by dbritto-dev** [reply]
@TkDodo done

**Reply by dbritto-dev** [reply]
@TkDodo @dai-shi 
<img width="1480" alt="image" src="https://github.com/pmndrs/zustand/assets/1697714/3485dae3-179a-4a16-abcf-acc7c58497aa">

**Reply by dbritto-dev** [reply]
<img width="1255" alt="image" src="https://github.com/pmndrs/zustand/assets/1697714/b4915c4e-a9ca-4f89-978f-f7d45d3ccff9">

**Reply by dai-shi** [reply]
On second thought, I don't think we need createContext in the legacy api. We should have a guide for context usage, and optionally introduce a third-party library.

**Reply by dai-shi** [reply]
Can you separate two topics? 1) updating `docs` content, and 2) introducing new `website`.

**Reply by dbritto-dev** [reply]
@dai-shi what you mean "separate two topics"?

**Reply by dai-shi** [reply]
separate discussions (or discussion threads) and separate PRs.

**Reply by dbritto-dev** [reply]
@dai-shi I got it, you mean updating docs content and then introduce the new website that use docs content. I don't think we can do that, since the new docs follow the new syntax supported by vitepress and wouldn't work on the current setup

**Reply by dbritto-dev** [reply]
@dai-shi but if we want to have separate discussions just to track the changes we can do that but eventually they would be just one PR

**Reply by dai-shi** [reply]
> I don't think we can do that, since the new docs follow the new syntax supported by vitepress and wouldn't work on the current setup

Isn't it just ignored? Or, can you avoid incompatible syntax for the start? The whole point of separating `docs` and `website` is to make `docs` agnostic to `website`.

**Reply by dai-shi** [reply]
And, `docs` should be browsable on GitHub too.

**Reply by dbritto-dev** [reply]
@dai-shi the docs would be browsable on GitHub, we don't have issues with that

**Reply by dbritto-dev** [reply]
@dai-shi but let me work on updating our current docs to follow the new structure to support the new website

**Reply by dbritto-dev** [reply]
@dai-shi I'm going to create two discussions and two PRs as you are suggesting
